mixer/test/e2e/e2e_cachebuster_test.go

```diff
@@ -0,0 +1,164 @@
+  name: checkInstance
+  namespace: istio-system
+spec:
+  stringPrimitive: "destination.service"
+---
+
+apiVersion: "config.istio.io/v1alpha2"
+kind: rule
+metadata:
+  name: rule1
+  namespace: istio-system
+spec:
+  match: destination.service == "echosrv1.istio.svc.cluster.local" && source.service == "foo"
+#  match: destination.service == "echosrv1.istio.svc.cluster.local" && request.headers["x-request-id"] == "foo"
+  actions:
+  - handler: fakeHandlerConfig.fakeHandler
+    instances:
+    - checkInstance.samplecheck
+
+---
+`
+)
+
+func TestCacheBuster(t *testing.T) {
+	tests := []testData{
+		{
+			name:      "CacheBuster",
+			cfg:       cacheBusterSvcCfg,
+			behaviors: []spyAdapter.AdapterBehavior{{Name: "fakeHandler"}},
+			templates: e2eTmpl.SupportedTmplInfo,
+			attrs:     map[string]interface{}{
+				"destination.service": "echosrv2.istio.svc.cluster.local",
+				"source.service": "foo",
+				"request.headers": map[string]interface{}{
+					"x-request-id": "foo",
+				},
+			},
+			validate: func(t *testing.T, err error, spyAdpts []*spyAdapter.Adapter) {
+				adptr := spyAdpts[0]
+
+				if adptr.HandlerData.HandleSampleCheckInstance != nil {
+					t.Fatalf("Unexpected instance encountered: %v", adptr.HandlerData.HandleSampleCheckInstance)
+				}
+			},
+			validateCheckResponse: func(t *testing.T, response *istio_mixer_v1.CheckResponse, err error) {
+				if err != nil {
+					t.Fatalf("Unexpected error: %v", err)
+				}
+
+				// Expect destination.service (and context.protocol) to be referenced, but not request.headers, or source.service)
+				expected := `
+&ReferencedAttributes{Words:[context.protocol destination.service],AttributeMatches:[{-1 ABSENCE  0} {-2 EXACT  0}],}
+`
+				actual := response.Precondition.ReferencedAttributes.String()
+				if strings.TrimSpace(expected) != strings.TrimSpace(actual) {
+					t.Fatalf("got: \n%v\nwanted:\n%v\n", actual, expected)
+				}
+			},
+		},
+	}
+	for i, tt := range tests {
+		t.Run(strconv.Itoa(i), func(t *testing.T) {
+			adapterInfos, spyAdapters := ConstructAdapterInfos(tt.behaviors)
+
+			args := testEnv.NewArgs()
+			args.APIPort = 0
+			args.MonitoringPort = 0
+			args.Templates = e2eTmpl.SupportedTmplInfo
+			args.Adapters = adapterInfos
+			var cerr error
+			if args.ConfigStore, cerr = storetest.SetupStoreForTest(cacheBusterGlobalCfg, tt.cfg); cerr != nil {
+				t.Fatal(cerr)
+			}
+
+			env, err := testEnv.New(args)
+			if err != nil {
+				t.Fatalf("fail to create mixer: %v", err)
+			}
+
+			env.Run()
+
+			defer closeHelper(env)
+
+			conn, err := grpc.Dial(env.Addr().String(), grpc.WithInsecure())
+			if err != nil {
+				t.Fatalf("Unable to connect to gRPC server: %v", err)
+			}
+
+			client := istio_mixer_v1.NewMixerClient(conn)
+			defer closeHelper(conn)
+
+			req := istio_mixer_v1.CheckRequest{
+				Attributes: getAttrBag(tt.attrs,
+					args.ConfigIdentityAttribute,
+					args.ConfigIdentityAttributeDomain),
+			}
+			tt.validate(t, err, spyAdapters)
```

Comment:
why are we validating not nil before the call ?

Reply:
The field may not be present for all cases.